### PR TITLE
support pass-in build configuration for image build

### PR DESCRIPTION
**Pull request by philipktlin**
allow to pass build configuration for image build; default value is "Release"

### CODE REVIEW

File: builds/misc/images.yaml [hunk 1]
```diff
@@ -31,9 +31,9 @@ jobs:
       - bash: 'cargo install --git https://github.com/arsing/cross.git --branch set-path'
         displayName: 'Install cross (fork with docker fix)'
 
-      - script: scripts/linux/buildBranch.sh -c Release --no-rocksdb-bin --os Unix --dotnet_runtime netcoreapp3.0
+      - script: scripts/linux/buildBranch.sh -c $(Build.Configuration) --no-rocksdb-bin --os Unix --dotnet_runtime netcoreapp3.0
         name: build
-        displayName: Build (release)
+        displayName: Build ($(Build.Configuration))
 
       - script: scripts/linux/buildDiagnostics.sh
```
Comment (damonbarry) [inline builds/misc/images.yaml]: The diagnostics tool is always built with release config (in `buildDiagnostics.sh` and `build-diagnostics.ps1`). Is that OK? Or do we also need to make that build configurable?

Comment (philipktlin) [reply builds/misc/images.yaml]: Update these script to take build configuration from parameter; you can verify from this build, https://msazure.visualstudio.com/One/_build/results?buildId=21356904.

File: edgelet/build/windows/build-diagnostics.ps1 [hunk 2]
```diff
@@ -4,12 +4,14 @@
  # Builds and publishes to target/publish/ the iotedge-diagnostics binary and its associated dockerfile
  #>
 
+param([string]$BuildConfiguration = "release")
```
Comment (damonbarry) [inline edgelet/build/windows/build-diagnostics.ps1]: Consider adding `[ValidateSet("debug", "release")]`

File: edgelet/build/windows/build-diagnostics.ps1 [hunk 3]
```diff
@@ -19,8 +21,8 @@ $env:NO_VALGRIND = 'true'
 
 $originalRustflags = $env:RUSTFLAGS
 $env:RUSTFLAGS += ' -C target-feature=+crt-static'
-Write-Host "$cargo build -p iotedge-diagnostics --release --manifest-path $ManifestPath"
-Invoke-Expression "$cargo build -p iotedge-diagnostics --release --manifest-path $ManifestPath"
+Write-Host "$cargo build -p iotedge-diagnostics --$BuildConfiguration --manifest-path $ManifestPath"
```
Comment (damonbarry) [inline edgelet/build/windows/build-diagnostics.ps1]: I don't think `--debug` is a valid argument to `cargo build`. You either include or omit `--release`.

File: scripts/linux/buildDiagnostics.sh [hunk 4]
```diff
@@ -5,6 +5,9 @@
 # It then publishes the binaries along with their corresponding dockerfiles to the publish directory,
 # so that buildImage.sh can build the container image.
 
+BuildConfiguration=${1:-release}
```
Comment (damonbarry) [inline scripts/linux/buildDiagnostics.sh]: In this script variables are in UPPERCASE_SNAKE format, not pascal.

File: scripts/linux/buildDiagnostics.sh [hunk 5]
```diff
@@ -16,14 +19,15 @@ cp -R $BUILD_REPOSITORY_LOCALPATH/edgelet/iotedge-diagnostics/docker $PUBLISH_FO
 
 cd "$BUILD_REPOSITORY_LOCALPATH/edgelet"
 
-cross build -p iotedge-diagnostics --release --target x86_64-unknown-linux-musl
-strip $BUILD_REPOSITORY_LOCALPATH/edgelet/target/x86_64-unknown-linux-musl/release/iotedge-diagnostics
-cp $BUILD_REPOSITORY_LOCALPATH/edgelet/target/x86_64-unknown-linux-musl/release/iotedge-diagnostics $PUBLISH_FOLDER/azureiotedge-diagnostics/docker/linux/amd64/
+BuildOption="--$BuildConfiguration"
```
Comment (damonbarry) [inline scripts/linux/buildDiagnostics.sh]: Same as above, the argument to `cargo build` is either `--release` or nothing.

File: builds/misc/images.yaml [hunk 6]
```diff
@@ -4,6 +4,14 @@ trigger:
     include:
       - master
 pr: none
+
+# Variable defined in VSTS
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply builds/misc/images.yaml]: removed.

File: edgelet/build/windows/build-diagnostics.ps1 (hunk 2 above)
Comment (philipktlin) [reply edgelet/build/windows/build-diagnostics.ps1]: added

File: scripts/linux/buildDiagnostics.sh (hunk 4 above)
Comment (philipktlin) [reply scripts/linux/buildDiagnostics.sh]: updated

File: edgelet/build/windows/build-diagnostics.ps1 [hunk 7]
```diff
@@ -4,11 +4,23 @@
...
+    [ValidateSet("debug", "release")]
+    [string]$BuildConfiguration = "release"
+)
+ 
 $ErrorActionPreference = 'Continue'
 
 . (Join-Path $PSScriptRoot 'util.ps1')
 
 Assert-Rust
+If ($BuildConfiguration.ToLower() == "release") {
```
Comment (arsing) [inline edgelet/build/windows/build-diagnostics.ps1]: `if ($BuildConfiguration -eq 'release') {`

PS does not have `==`, and `-eq` is already case-insensitive.

File: edgelet/build/windows/build-diagnostics.ps1 [hunk 8]
```diff
@@ -4,11 +4,23 @@
  # Builds and publishes to target/publish/ the iotedge-diagnostics binary and its associated dockerfile
  #>
 
+param(
+    [ValidateSet("debug", "release")]
```
Comment (arsing) [inline edgelet/build/windows/build-diagnostics.ps1]: Single quotes for all strings that don't have interpolation.

File: edgelet/build/windows/build-diagnostics.ps1 [hunk 9]
```diff
@@ -4,11 +4,23 @@
...
+    [string]$BuildConfiguration = "release"
+)
+ 
 $ErrorActionPreference = 'Continue'
 
 . (Join-Path $PSScriptRoot 'util.ps1')
 
 Assert-Rust
+If ($BuildConfiguration.ToLower() == "release") {
+    $BuildConfiguration="release"
```
Comment (arsing) [inline edgelet/build/windows/build-diagnostics.ps1]: Spaces around `=`

File: scripts/linux/buildDiagnostics.sh [hunk 10]
```diff
@@ -5,6 +5,16 @@
 # It then publishes the binaries along with their corresponding dockerfiles to the publish directory,
 # so that buildImage.sh can build the container image.
 
+BUILD_CONFIGURATION=${1:-release}
+
+if [[ "${BUILD_CONFIGURATION,,}" == "release" ]]; then
```
Comment (arsing) [inline scripts/linux/buildDiagnostics.sh]: Single quotes around strings that don't have interpolation. (`'release'`, etc)

**Comment by arsing** [comment]
(Please ensure Build Images job passes before merging into master.)

**Comment by philipktlin** [comment]
Tested.
Debug: https://msazure.visualstudio.com/One/_build/results?buildId=21359408
Release: https://msazure.visualstudio.com/One/_build/results?buildId=21359393

**Pull request merged**